examples/7GUIs/circle_drawer.rs

```diff
@@ -0,0 +1,358 @@
+    }
+}
+
+impl Model for CircleDrawerData {
+    fn event(&mut self, cx: &mut EventContext, event: &mut Event) {
+        if let Some(event) = event.take() {
+            match event {
+                CircleDrawerEvent::AddCircle(x, y) => {
+                    let circle = Circle { x, y, r: 26.0 };
+                    self.circles_data.selected = Some(self.circles_data.circles.len());
+                    self.circles_data.circles.push(circle);
+                    self.undo_list.push(UndoRedoAction::Circle(circle));
+                    self.redo_list.clear(); // we need to clear redo list in each new event
+                }
+                CircleDrawerEvent::TrySelectCircle(x, y) => {
+                    if !(self.dialog_open || self.menu_open) {
+                        self.update_selected(x, y)
+                    }
+                }
+                CircleDrawerEvent::ChangeRadius(r) => {
+                    if let Some(idx) = self.circles_data.selected {
+                        self.circles_data.circles[idx].r = r;
+                    }
+                }
+                CircleDrawerEvent::Undo => {
+                    let last = self.undo_list.remove(self.undo_list.len() - 1);
+
+                    match last {
+                        UndoRedoAction::Circle(_) => {
+                            self.redo_list.push(last);
+                            self.circles_data.circles.pop(); // remove the last circle
+                        }
+                        UndoRedoAction::RadiusChange(idx, r) => {
+                            self.redo_list.push(UndoRedoAction::RadiusChange(
+                                idx,
+                                self.circles_data.circles[idx].r, // store the current radius in redo list
+                            ));
+                            self.circles_data.circles[idx].r = r // update the radius to the old one
+                        }
+                    }
+                }
+                CircleDrawerEvent::Redo => {
+                    let last = self.redo_list.remove(self.redo_list.len() - 1);
+
+                    match last {
+                        UndoRedoAction::Circle(c) => {
+                            self.undo_list.push(last);
+                            self.circles_data.circles.push(c);
+                        }
+                        UndoRedoAction::RadiusChange(idx, r) => {
+                            self.undo_list.push(UndoRedoAction::RadiusChange(
+                                idx,
+                                self.circles_data.circles[idx].r, // store the current radius in undo list
+                            ));
+                            self.circles_data.circles[idx].r = r // update the radius
+                        }
+                    }
+                }
+                CircleDrawerEvent::ToggleRightMenu => {
+                    let x = cx.mouse().cursorx;
+                    let y = cx.mouse().cursory;
```

Comment:
It probably doesn't matter here but it might be better to get the pressed position of the mouse button rather than the current position. You can do this with `cx.mouse().left.pos_down`.

Reply:
My main concern would be that the mouse might move before this event is processed, causing the menu to appear not where the click occurred but where the mouse currently is. It's probably unlikely to happen, but using the click position rather than current position would fix this. Is it left click or right click that opens the menu? You can also do `cx.mouse().right.pos_down` to get the mouse position when the right button was last pressed.